packages/angular/ssr/src/routes/route-config.ts

```diff
@@ -0,0 +1,146 @@
+  CSR,
+
+  /** Static Site Generation (SSG) mode, where content is pre-rendered at build time and served as static files. */
+  SSG,
+}
+
+/**
+ * Fallback strategies for Static Site Generation (SSG) routes.
+ * @developerPreview
+ */
+export enum SSGFallback {
+  /** Use Server-Side Rendering (SSR) as the fallback for this route. */
+  SSR,
+
+  /** Use Client-Side Rendering (CSR) as the fallback for this route. */
+  CSR,
+
+  /** No fallback; Angular will not handle the response if the path is not pre-rendered. */
+  None,
+}
+
+/**
+ * Common interface for server routes, providing shared properties.
+ */
+export interface ServerRouteCommon {
+  /** The path associated with this route. */
+  path: string;
+
+  /** Optional additional headers to include in the response for this route. */
+  headers?: Record<string, string>;
+
+  /** Optional status code to return for this route. */
+  status?: number;
+}
+
+/**
+ * A server route that uses AppShell rendering mode.
+ */
+export interface ServerRouteAppShell extends Omit<ServerRouteCommon, 'headers' | 'status'> {
+  /** Specifies that the route uses AppShell rendering mode. */
+  renderMode: RenderMode.AppShell;
+}
+
+/**
+ * A server route that uses Client-Side Rendering (CSR) mode.
+ */
+export interface ServerRouteCSR extends ServerRouteCommon {
+  /** Specifies that the route uses Client-Side Rendering (CSR) mode. */
+  renderMode: RenderMode.CSR;
+}
+
+/**
+ * A server route that uses Static Site Generation (SSG) mode.
+ */
+export interface ServerRouteSSG extends Omit<ServerRouteCommon, 'status'> {
+  /** Specifies that the route uses Static Site Generation (SSG) mode. */
+  renderMode: RenderMode.SSG;
+
+  /**
+   * Optional fallback strategy to use if the SSG path is not pre-rendered.
+   * Defaults to `SSGFallback.SSR` if not provided.
+   */
+  fallback?: SSGFallback;
+  /**
+   * A function that returns a Promise resolving to an array of objects, each representing a route path with URL parameters.
+   * This function runs in the injector context, allowing access to Angular services and dependencies.
+   *
+   * @returns A Promise resolving to an array where each element is an object with string keys (representing URL parameter names)
+   * and string values (representing the corresponding values for those parameters in the route path).
+   *
+   * @example
+   * ```typescript
+   * export const serverRouteConfig: ServerRoutes[] = [
+   *   {
+   *     path: '/product/:id',
+   *     remderMode: RenderMode.SSG,
+   *     async getPrerenderPaths() {
+   *       const productService = inject(ProductService);
+   *       const ids = await productService.getIds(); // Assuming this returns ['1', '2', '3']
+   *
+   *       return ids.map(id => ({ id })); // Generates paths like: [{ id: '1' }, { id: '2' }, { id: '3' }]
+   *     },
+   *   },
+   * ];
+   * ```
+   */
+  getPrerenderPaths?: () => Promise<Record<string, string>[]>;
```

Comment:
Question: I don't suppose it would be possible to make `getPrerenderPaths` type dependent on `path` such that it could infer / validate all the required params are given?

I [made an attempt](https://www.typescriptlang.org/play/?#code/JYOwLgpgTgZghgYwgAgEoHsCukA8AFOMAC2QgA9IQATAZ2RrClAHMA+ZAbwChlfkAHQkQBcyAsQDcPPswhg8UaBGrRxRGgAoAlKIJQ4AWxo4AIlHT8AMhDhUWAZQA2cGkRwA5dAEkQMaPiFWIKkAXy4uMABPfhQzC2tbB2dXAOJSCmVaekYWdgBeMSF0yiyAAwB6ABIOUD8oNAgGENLkAH4GhmRREAgAN2gpCOiUPUNjNWLMugYmEDZkAonyErpS4Wra6EL9AxCqmt8t1EawZuleduOEdCgqAJ2AGmzZ+YAybbGcY4ZWc67CtLLKbIUrVGYsPYbQ71b6nUp-PhtD5GL4nX6IxG6IpA6irdYHOrIs4YklIq43O6jAxPcFzdGkzGcEKDcrlZAAFSIwDoAHcbgBrGgAOiGMQ5JwWyOMcSsNjscycLjcAHJysIYOh0OVIAw1QAjOBQbUnfVwABeyuC4VZHK5dCo6EaIGVYFaXGuIE65mwjUlAG0-twGYJiKJVerNcbdcIDUadWBTRaHgiZHIFEoVFA1DRREGGXwNegw4Xlcn83xY2HY6WUySDWaq+aawyQmW+K2uABdeiEbkwYC+jA+nC0th+zsSIA) at this, but it seems that making two properties of an object dependent on each other is not as straightforward as I expected. Maybe I'm missing something?

This might be too cute to be worth doing anyways, but something we should at least consider.

Reply:
I'll experiment a bit tomorrow. However, in your case, the issue is caused by `satisfies Route<string>[];`, which broadens the types.

Reply:
[This](https://www.typescriptlang.org/play/?noUncheckedIndexedAccess=true&noUnusedParameters=true&exactOptionalPropertyTypes=true&noPropertyAccessFromIndexSignature=true#code/JYOwLgpgTgZghgYwgAgEoHsCukA8AFOMAC2QgA9IQATAZ2RrClAHMAaZPKaCa6A4mgD5kAbwBQyScgAOhIgC4OcgNwSpzCGE7deUfkRoAKAJSLO6ALbAaEfFy679Q1QF8xYsAE9pKfQHVgYiwwABkIOCoWAGUAGzgaInw5UgoeWnpGFmEAXjVJfRTKdIADAHoAEhFQGGg0CAYXYuQAfjqGZDMVdy8fJWICKDgLTWgaJOJCtLoGJhBmHLy+knIiumL5SuragaGXCqqQGqg2sEbF1pFkAG0AaWRQJUGLAF1FGZZkF2QAMiWd4cgUDGqHqYEEi06ExWU2QxUq7zme02h1qIIaxXOfzgTxGQJwaLBEKWk2oaw2ByOj12GKkLVE1zuD3+rwys2YnyJIhcqjEMEwIAQYGA6BAyAQXEIEAw2FsBWhpNZWUMiygwQgimluH07H0-1xY38gSIwTCEWicQS4yIghtYlMaDVVp1cj1gINcgCQWwpsic1i8US+htwnEtK4YEwUFFqplrncCBF7RjkDo2Wui3F4UgmoghlDtKksmIigARKV5DB0OhSimwOWAEbYmughtwABeJdYi1p8U8AuQGi09jSfDkRmMom7Bck4cjovz08XyEr6FLK87U6XkkbUFLO43W8Xjbbe-bB8Pn1US5cXdpLmMt6kmclObzm5kclL5eAVGbDHP069v2g7aA4VCjgIJiTheyCzlG0EwVIP5rlWAGISsgzmNIigAOQ4cgpSlMgVDoMgIDoGApBQKqUDvneV6Ljeiz3o+kjPtmapvkuRYKMgZYVlWf5gGhPY0H2CADpooEjnoY5QQuW5wfOdGLshfHrqxW7cu+TF3g+YjPMg8RiomYCqEAA) is what I managed to achieve

Reply:
Interesting, I was hoping this could be done without changing the syntax but I don't see a better way than what you have.

In terms of value, `ng serve` would not run `getPrerenderPaths` right? If a developer doesn't go out of their way to check that `ng build` passes, they might not notice such an issue until CI runs. I can see some value in getting type checking here.

That said, I'm not totally sure it's worth a no-op `createRoute` function. Is this idea worth considering in the design or something we should just drop for now and come back to later if/when the cost/benefit is more clear?